[css-anchor-position-1] Scroll adjustment not applied for position:fixed anchored boxes

anttijk opened the pull request:
#### de4580fe7eeedb0b8582ce8ea85919b788e1d58b
<pre>
[css-anchor-position-1] Scroll adjustment not applied for position:fixed anchored boxes
<a href="https://bugs.webkit.org/show_bug.cgi?id=295002">https://bugs.webkit.org/show_bug.cgi?id=295002</a>
<a href="https://rdar.apple.com/154348421">rdar://154348421</a>

Reviewed by Simon Fraser.

Compute scroll adjustment correctly for fixed positioned anchored boxes.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::addViewportConstrainedObject):
(WebCore::LocalFrameView::removeViewportConstrainedObject):
(WebCore::LocalFrameView::hasAnchorPositionedViewportConstrainedObjects const):

See if any of the viewport constrained objects are anchor positioned.

(WebCore::LocalFrameView::clearCachedHasAnchorPositionedViewportConstrainedObjects):
(WebCore::LocalFrameView::shouldUpdateCompositingLayersAfterScrolling const):

We need to update compositing layers on every frame if there are fixed anchor positioned boxes.

(WebCore::LocalFrameView::updateAnchorPositionedAfterScroll):

We need to update anchor positions after document scroll too.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollPositionChanged):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::updateAnchorPositionedAfterScroll):

Factor into a separate virtual function, invoked from ScrollableArea::scrollPositionChanged.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleWillChange):

Clear the cached fixed-with-anchor bit if anchor positioned style changes.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPositionsAfterOverflowScroll):
(WebCore::RenderLayer::updateLayerPositionsAfterDocumentScroll):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):

Move to RenderLayer::updateLayerPositionsAfterOverflowScroll.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::scrollOffsetFromAnchor):

Take fixed position into account when computing scroll offset.

(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

Use defaultAnchorForBox() to figure out if we need scroll compensation.

(WebCore::Style::AnchorPositionEvaluator::updatePositionsAfterScroll):
(WebCore::Style::AnchorPositionEvaluator::isAnchorPositioned):
(WebCore::Style::scrollOffsetFromAncestorContainer): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updateAfterOverflowScroll): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Canonical link: <a href="https://commits.webkit.org/296811@main">https://commits.webkit.org/296811@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e52b786be56ab506342d4909281a6db3075d3e67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109660 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29318 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19747 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115680 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59894 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83333 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23898 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98759 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63792 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23279 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16903 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59475 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93273 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16941 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118473 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36699 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27177 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92340 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37072 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92161 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23480 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37117 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14865 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32527 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36593 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42064 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36253 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39596 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37963 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1006,6 +1006,9 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
             newStyleSlowScroll = false;
     }
+    // FIXME: Scrolling tree support for anchor positioning.
+    if (newStyle.position() == PositionType::Fixed && Style::AnchorPositionEvaluator::isAnchorPositioned(newStyle))
```
What about sticky positioning?

anttijk replied:
Sticky can't be anchor positioned, only fixed and absolute.

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -1176,7 +1179,7 @@ void RenderElement::willBeDestroyed()
 #endif
-    if (m_style.hasAnyFixedBackground() && !settings().fixedBackgroundsPaintRelativeToDocument())
+    if ((m_style.hasAnyFixedBackground() && !settings().fixedBackgroundsPaintRelativeToDocument()) || (m_style.position() == PositionType::Fixed && Style::AnchorPositionEvaluator::isAnchorPositioned(m_style)))
         view().protectedFrameView()->removeSlowRepaintObject(*this);
```
It's not ideal to use the "slow repaint" logic for this; this is about needing to layout after scroll, not about painting. This is more like the `setViewportConstrainedObjectsNeedLayout` code path.

anttijk replied:
This is about needing compositing update for every frame.

anttijk replied:
Basically LocalFrameView::shouldUpdateCompositingLayersAfterScrolling() needs to return true in this case.

anttijk replied:
Did a version using viewport constrained objects.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -6012,6 +6013,9 @@ void RenderLayerCompositor::updateSynchronousScrollingNodes()
             rootHasSlowRepaintObjects = true;
+        } else if (Style::AnchorPositionEvaluator::isAnchorPositioned(renderer.style())) {
+            LOG_WITH_STREAM(Scrolling, stream << "RenderLayerCompositor::updateSynchronousScrollingNodes - root node slow-scrolling because of a fixed anchored element");
+            rootHasSlowRepaintObjects = true;
```
Again, mixing layout with repainting.

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -286,15 +286,24 @@ static LayoutSize offsetFromAncestorContainer(const RenderElement& descendantCon
 
     auto offset = LayoutSize { };
-    for (auto* ancestor = descendant.container(); ancestor && ancestor != &ancestorContainer; ancestor = ancestor->container()) {
+    bool isFixedAnchor = anchor.isFixedPositioned();
```
Need to consider sticky.

anttijk replied:
See above.

smfr commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -286,15 +286,24 @@ static LayoutSize offsetFromAncestorContainer(const RenderElement& descendantCon
-    for (auto* ancestor = descendant.container(); ancestor && ancestor != &ancestorContainer; ancestor = ancestor->container()) {
+    bool isFixedAnchor = anchor.isFixedPositioned();
+
+    for (auto* ancestor = anchor.container(); ancestor && ancestor != containingBlock; ancestor = ancestor->container()) {
```
CheckedPtr?

anttijk committed:
[css-anchor-position-1] Scroll adjustment not applied for position:fixed anchored boxes
https://bugs.webkit.org/show_bug.cgi?id=295002
rdar://154348421

Reviewed by Simon Fraser.

Compute scroll adjustment correctly for fixed positioned anchored boxes.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-008-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-010-expected.txt:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::addViewportConstrainedObject):
(WebCore::LocalFrameView::removeViewportConstrainedObject):
(WebCore::LocalFrameView::hasAnchorPositionedViewportConstrainedObjects const):

See if any of the viewport constrained objects are anchor positioned.

(WebCore::LocalFrameView::clearCachedHasAnchorPositionedViewportConstrainedObjects):
(WebCore::LocalFrameView::shouldUpdateCompositingLayersAfterScrolling const):

We need to update compositing layers on every frame if there are fixed anchor positioned boxes.

(WebCore::LocalFrameView::updateAnchorPositionedAfterScroll):

We need to update anchor positions after document scroll too.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollPositionChanged):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::updateAnchorPositionedAfterScroll):

Factor into a separate virtual function, invoked from ScrollableArea::scrollPositionChanged.

* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleWillChange):

Clear the cached fixed-with-anchor bit if anchor positioned style changes.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateLayerPositionsAfterOverflowScroll):
(WebCore::RenderLayer::updateLayerPositionsAfterDocumentScroll):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):

Move to RenderLayer::updateLayerPositionsAfterOverflowScroll.

* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::scrollOffsetFromAnchor):

Take fixed position into account when computing scroll offset.

(WebCore::Style::AnchorPositionEvaluator::updateSnapshottedScrollOffsets):

Use defaultAnchorForBox() to figure out if we need scroll compensation.

(WebCore::Style::AnchorPositionEvaluator::updatePositionsAfterScroll):
(WebCore::Style::AnchorPositionEvaluator::isAnchorPositioned):
(WebCore::Style::scrollOffsetFromAncestorContainer): Deleted.
(WebCore::Style::AnchorPositionEvaluator::updateAfterOverflowScroll): Deleted.
* Source/WebCore/style/AnchorPositionEvaluator.h:

Canonical link: https://commits.webkit.org/296811@main
